Marlin/src/libs/hex_print_routines.cpp

```diff
@@ -21,18 +21,14 @@
  */
 
 #include "../inc/MarlinConfig.h"
+#include "../gcode/parser.h"
 
 #if ENABLED(AUTO_BED_LEVELING_UBL) || ENABLED(M100_FREE_MEMORY_WATCHER) || ENABLED(DEBUG_GCODE_PARSER)
 
   #include "hex_print_routines.h"
 
-  #ifdef CPU_32_BIT
-    constexpr int byte_start = 0;
-    static char _hex[] = "0x0000";
-  #else
-    constexpr int byte_start = 4;
-    static char _hex[] = "0x00000000";
-  #endif
+  constexpr int byte_start = 4;
+  static char _hex[] = "0x00000000";
```

Comment:
This setup is needed to correctly show "0x" in front of hex addresses, whether they are 16 bit or 32-bit. By always making `byte_start` 4 breaks the display of addresses.

I should go over this code once more to make sure the offsets are sound. I chose 0 and 4 just so the first offset would be 0 and the second positive. The full code should probably read like this:
```cpp
  #ifdef CPU_32_BIT
    constexpr int byte_start = 0;
    static char _hex[] = "0x0000";
  #else
    constexpr int byte_start = 4;
    static char _hex[] = "0x00000000";
  #endif

  char* hex_byte(const uint8_t b) {
    _hex[byte_start + 4] = hex_nybble(b >> 4);
    _hex[byte_start + 5] = hex_nybble(b);
    return &_hex[byte_start + 4];
  }
 
  char* hex_word(const uint16_t w) {
    _hex[byte_start + 2] = hex_nybble(w >> 12);
    _hex[byte_start + 3] = hex_nybble(w >> 8);
    _hex[byte_start + 4] = hex_nybble(w >> 4);
    _hex[byte_start + 5] = hex_nybble(w);
    return &_hex[byte_start + 2];
  }
 
  #ifdef CPU_32_BIT
    char* hex_long(const uint32_t l) {
      _hex[byte_start - 2] = hex_nybble(l >> 28);
      _hex[byte_start - 1] = hex_nybble(l >> 24);
      _hex[byte_start + 0] = hex_nybble(l >> 20);
      _hex[byte_start + 1] = hex_nybble(l >> 16);
      (void)hex_word((uint16_t)(l & 0xFFFF));
      return &_hex[byte_start - 2];
    }
  #endif
```

Reply:
That makes a lot more sense, as the original code started like this:

```
  #ifdef CPU_32_BIT
    constexpr int byte_start = 0;
    static char _hex[] = "0x0000";
  #else
    constexpr int byte_start = 4;
    static char _hex[] = "0x00000000";
  #endif
```

As such, the `_hex` arrays and `byte_start` values were reversed.

Reply:
There is only a problem I see there, in 32bit CPUs Words and longs are both 32bits. 
With the defines above, we get 16bit words, and 32bit longs.
This is the way I did it, and works fine in my STM32F1.

`
  #ifdef CPU_32_BIT
    constexpr int byte_start = 8;
    static char _hex[11] = "0x00000000";
  #else
    constexpr int byte_start = 4;
    static char _hex[7] = "0x0000";
  #endif

  char* hex_byte(const uint8_t b) {
    _hex[byte_start] = hex_nybble(b >> 4);
    _hex[byte_start + 1] = hex_nybble(b);
    return &_hex[byte_start];
  }


  #ifdef CPU_32_BIT
    char* hex_word(const uint32_t w) {
      _hex[2] = hex_nybble(w >> 28);
      _hex[3] = hex_nybble(w >> 24);
      _hex[4] = hex_nybble(w >> 20);
      _hex[5] = hex_nybble(w >> 16);
      _hex[6] = hex_nybble(w >> 12);
      _hex[7] = hex_nybble(w >> 8);
      _hex[8] = hex_nybble(w >> 4);
      _hex[9] = hex_nybble(w);
      return &_hex[2];
  }
  #else
    char* hex_word(const uint16_t w) {
      _hex[byte_start + 2] = hex_nybble(w >> 12);
      _hex[byte_start + 3] = hex_nybble(w >> 8);
      _hex[byte_start + 4] = hex_nybble(w >> 4);
      _hex[byte_start + 5] = hex_nybble(w);
      return &_hex[2];
    }
  #endif

  char* hex_address(const void * const w) {
      (void)hex_word((ptr_int_t)w);
    return _hex;
  }

  void print_hex_nybble(const uint8_t n)       { SERIAL_CHAR(hex_nybble(n));  }
  void print_hex_byte(const uint8_t b)         { SERIAL_ECHO(hex_byte(b));    }
  void print_hex_word(const ptr_int_t w)        { SERIAL_ECHO(hex_word(w));    }
  void print_hex_address(const void * const w) { SERIAL_ECHO(hex_address(w)); }

#endif`

Reply:
@victorpv the length of a word is debatable, with some platforms considering it to be the native length of a pointer, and others fixing it at 16 bits. I think @thinkyhead's definition of a word as 16 bits is convenient from the perspective that it's consistent across both the 8-bit AVR platform and the new 32-bit platforms.

Reply:
"word" is not a C native type, nor is it in stdint.h, so it is always user defined. However, Arduino define it as 16 or 32 bits depending on platform. https://www.arduino.cc/en/Reference/Word

Reply:
Following on Bobx comment, I think we are going to fix it to 16bits, why don't we call it short? and the 32bit long as in this latest version. That would prevent any confusion.
` inline void _hex_short(const uint16_t w) {`